sycl/doc/SharedLibraries.md

```diff
@@ -0,0 +1,379 @@
+  situation leads to undefined behaviour. For more details refer to
+  [ODR violations](#ODR-violations) section.
+
+Each combination of symbols is assigned with an Id number - symbol set Id.
+A combination of symbols can exist in different formats (i.e. SPIR-V/AOT
+compiled binary and etc).
+`m_DeviceImages` maps an Id number to an array with device images which represent
+the same combination of symbols in different formats.
+`m_SymbolSets` contains mapping from symbol name to symbol set Id for each OS
+module (.exe/.so/.dll).
+`std::unordered_map` allows to search and access its elements with constant-time
+complexity.
+
+Before compilation of device image to execute a kernel RT checks if the image
+contains any import information in its properies and if it does, then RT
+performs device images collection in order to resolve dependencies.
+
+Ids of all needed symbol sets are found. This is done by iterating through
+`m_SymbolSets` map, i.e. iterating through all available OS modules without
+predefined order and searching for first unresolved symbol in list of imports
+set of target device image. Once device image that contains first symbol is
+met, remaining exported symbols are checked in found image and if
+they match some imported symbols then these matched symbols will be marked as
+resolved. The procedure repeats until all imported symbols are resolved.
+For each found symbol set Id program cache is checked in case if
+necessary set of `SYCL_EXTERNAL` functions has been compiled and if it is true,
+then compiled device image will be re-used for linking.
+Otherwise device image containing required symbols set will be compiled and
+stored in cache.
+
+#### Program caching
+
+Existing support for device code caching is re-used to cache programs created
+from device images with SYCL external functions and linked device images with
+imports information.
+
+##### In-memory cache
+
+Programs that contain only `SYCL_EXTERNAL` functions will be cached only in
+compiled state, so they can be linked with other programs during dependency
+resolution.
+
+The existing mechanism of caching is not changed for programs with
+imports information. They are stored in cache after they compiled and linked
+with programs that provide their dependencies. To identify linked programs
+Id of "main" set of symbols (i.e. the one which actually contain kernels) will
+be used.
+
+##### Persistent cache
+
+The documented approach to persistent cache needs to be expanded in presence
+of dynamic linking support. One of the identifiers for built image hash is
+hash made out of device image used as input for the JIT compilation.
+In case when "main" image have imports information, device image hash should be
+created from all device images that are necessary to build it, i.e. hash out
+of "main" device image and set of 'SYCL_EXTERNAL'-only images that define all
+symbols imported by "main device image.
```

Comment:
Current approach assume storing final binary after linking it means:
 - duplication of library images in every `main` program;
 - on change of library all `main` images in persistent cache are invalidated and JIT is initiated.
May be it is reasonable to store library `images` in cache separately from `main` image? It can be done both for in-memory and persistent or only for persistent cache depending on `weight` of link operation for binary programs.

Reply:
> May be it is reasonable to store library images in cache separately from main image? It can be done both for in-memory and persistent or only for persistent cache depending on weight of link operation for binary programs.

Do you mean compile to binary "main" image and "library" image and store them on disk separately? It sounds reasonable (the real dynamic libraries actually work this way), however I don't think we have devices that support linking of native device binaries yet, so right now It doesn't seem possible.